src/function/window/window_token_tree.cpp

```diff
@@ -0,0 +1,123 @@
+#include "duckdb/function/window/window_token_tree.hpp"
+
+namespace duckdb {
+
+class WindowTokenTreeLocalState : public WindowMergeSortTreeLocalState {
+public:
+	explicit WindowTokenTreeLocalState(WindowTokenTree &token_tree)
+	    : WindowMergeSortTreeLocalState(token_tree), token_tree(token_tree) {
+	}
+	//! Process sorted leaf data
+	void BuildLeaves() override;
+
+	WindowTokenTree &token_tree;
+};
+
+void WindowTokenTreeLocalState::BuildLeaves() {
+	auto &global_sort = *token_tree.global_sort;
+	if (global_sort.sorted_blocks.empty()) {
+		return;
+	}
+
+	//	Scan the sort keys and note deltas
+	SBIterator curr(global_sort, ExpressionType::COMPARE_LESSTHAN);
+	SBIterator prev(global_sort, ExpressionType::COMPARE_LESSTHAN);
+	const auto &sort_layout = global_sort.sort_layout;
+
+	const auto block_begin = token_tree.block_starts.at(build_task);
+	const auto block_end = token_tree.block_starts.at(build_task + 1);
+	auto &deltas = token_tree.deltas;
+	if (!block_begin) {
+		// First block, so set up initial delta
+		deltas[0] = 0;
+	} else {
+		// Move to the to end of the previous block
+		// so we can record the comparison result for the first row
+		curr.SetIndex(block_begin - 1);
+		prev.SetIndex(block_begin - 1);
+	}
+
+	for (++curr; curr.GetIndex() < block_end; ++curr, ++prev) {
+		int lt = 0;
+		if (sort_layout.all_constant) {
+			lt = FastMemcmp(prev.entry_ptr, curr.entry_ptr, sort_layout.comparison_size);
+		} else {
+			lt = Comparators::CompareTuple(prev.scan, curr.scan, prev.entry_ptr, curr.entry_ptr, sort_layout,
+			                               prev.external);
+		}
+
+		deltas[curr.GetIndex()] = (lt != 0);
+	}
+}
+
+idx_t WindowTokenTree::MeasurePayloadBlocks() {
+	const auto count = WindowMergeSortTree::MeasurePayloadBlocks();
+
+	deltas.resize(count);
+
+	return count;
+}
+
+template <typename T>
+static void BuildTokens(WindowTokenTree &token_tree, vector<T> &tokens) {
+	PayloadScanner scanner(*token_tree.global_sort);
+	DataChunk payload_chunk;
+	payload_chunk.Initialize(token_tree.context, token_tree.global_sort->payload_layout.GetTypes());
+	const T *row_idx = nullptr;
+	idx_t i = 0;
+
+	T token = 0;
+	for (auto &d : token_tree.deltas) {
+		if (i >= payload_chunk.size()) {
+			payload_chunk.Reset();
+			scanner.Scan(payload_chunk);
+			if (!payload_chunk.size()) {
+				break;
+			}
+			row_idx = FlatVector::GetData<T>(payload_chunk.data[0]);
+			i = 0;
+		}
+
+		token += d;
+		tokens[row_idx[i++]] = token;
+	}
+}
+
+unique_ptr<WindowAggregatorState> WindowTokenTree::GetLocalState() {
+	return make_uniq<WindowTokenTreeLocalState>(*this);
+}
+
+void WindowTokenTree::CleanupSort() {
+	//	Convert the deltas to tokens
+	if (mst64) {
```

Comment:
Is it worth converting the `if`/`else` control flow to an `enum`? Would we consider extending this with thinner(`uint16_t`)/wider(`uhugeint_t`) types in the future? If not, I'm fine with keeping this as it is.

Reply:
Not in the short term, but if I do, then yes we should switch off the width of the physical type.